enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,424 @@
+reconciling the machine.
+
+##### Reconciliation
+
+When a Hook Implementing Controller updates the Machine, reconciliation will be
+triggered, and the Machine will continue reconciling as normal, unless another
+hook is still present; there is no need to 'fail' the reconciliation to
+enforce requeuing.
+
+When all hooks for a given lifecycle-point are removed, reconciliation
+will continue as normal.
+
+##### Hook failure
+
+The machine controller should not timeout or otherwise consider the lifecycle
+hook as 'failed.'  Only the Hook Implementing Controller may decide to remove a
+particular lifecycle hook to allow the machine controller to progress past the
+corresponding lifecycle-point.
+
+##### Hook ordering
+
+The machine controller will not attempt to enforce any ordering of hooks.
+No ordering should be expected by the machine controller.
+
+Hook Implementing Controllers may choose to provide a mechanism to allow
+ordering amongst themselves via whatever means HICs determine.
+Examples could be using CRDs external to the Machine API, gRPC communications,
+or additional annotations on the Machine or other objects.
+
+#### Hook Implementing Controller Design
+
+Hook Implementing Controller is the component that manages a particular
+lifecycle hook.
+
+##### Hook Implementing Controllers must
+
+* Watch Machine objects and determine when an appropriate action must be taken.
+* After completing the desired hook action, remove the hook annotation.
+
+##### Hook Implementing Controllers may
+
+* Watch Machine objects and add a hook annotation as desired by the cluster
+administrator.
+* Coordinate with other Hook Implementing Controllers through any means
+possible, such as using common annotations, CRDs, etc. For example, one hook
+controller could set an annotation indicating it has finished its work, and
+another hook controller could wait for the presence of the annotation before
+proceeding.
+
+#### Determining when to take action
+
+A Hook Implementing Controller should watch Machines and determine when is the
+best time to take action.
+
+For example, if an HIC manages a lifecycle hook at the pre-drain lifecycle-point,
+then that controller should take action immediately after a Machine has a
+DeletionTimestamp or enters the "Deleting" phase.
+
+Fine-tuned coordination is not possible at this time; eg, it's not
+possible to execute a pre-terminate hook only after a node has been drained.
+This is reserved for future work.
+
+##### Failure Mode
+
+It is entirely up to the Hook Implementing Controller to determine when it is
+prudent to remove a particular lifecycle hook. Some controllers may want to
+'give up' after a certain time period, and others may want to block indefinitely.
+Cluster operators should consider the characteristics of each controller before
+utilizing them in their clusters.
+
+### Future Work
+
+#### Conditions
+
+To ensure that users have visibility into why a Machine has not been removed yet,
```

Comment:
Visibility seems logically required before we can support this feature in a release.  Why was this work deferred as future?